### ISSUE TITLE
Table rendering error

### ISSUE THREAD

**Issue by eduardomozart** [dokuwiki/dokuwiki#4051]
Hello,
When rendering the following table, the "Binário" section isn't shown. The markdown syntax seems correct, so I believe it's a DokuWiki issue. Tried on another instance with no plugins but it still isn't being rendered as it should. I'm running DokuWiki Release 2023-04-04a "Jack Jackrum".

```
====== Arquitetura de Computadores ======

===== Sistemas de Numeração =====

^  Decimal  ^  Binário  ^  Octal  ^  Hexadecimal  ^
|     0     |      0    |    0    |       0       |
|     1     |      1    |    1    |       1       |
|     2     |     10    |    2    |       2       |
|     3     |     11    |    3    |       3       |
|     4     |    100    |    4    |       4       |
|     5     |    101    |    5    |       5       |
|     6     |    110    |    6    |       6       |
|     7     |    111    |    7    |       7       |
|     8     |   1000    |   10    |       8       |
|     9     |   1001    |   11    |       9       |
|     10    |   1010    |   12    |       A       |
|     11    |   1011    |   13    |       B       |
|     12    |   1100    |   14    |       C       |
|     13    |   1101    |   15    |       D       |
|     14    |   1110    |   16    |       E       |
|     15    |   1111    |   17    |       F       |
|     16    |  10000    |   20    |      10       |
|     17    |  10001    |   21    |      11       |
|     18    |  10010    |   22    |      12       |
|     19    |  10011    |   23    |      13       |
|     20    |  10100    |   24    |      14       |
   
==== Binário ====
```

**Comment by fiwswe** [comment]
I don't know exactly why this happens, but in your example the line above the heading contains 3 spaces. If you delete at least 2 of the 3 spaces in the line directly above `==== Binário ====` then it works fine.

Actually a more reduced example looks like this:
```
| 1 |
  
==== X ====
```
Note the 2 spaces in the line before the `==== X ====`.

I think you always need to terminate a table with an empty line (i.e. no or less than 2 spaces or tabs on that line).
I suspect the cause is: https://github.com/dokuwiki/dokuwiki/blob/bac51761dc7fec256b7e3dbfe1205fbf1aad9a16/inc/Parsing/ParserMode/Table.php#L35
But I have no idea what the original intention of this line was. So simply removing it may not be an option and may break other things? Just terminate tables with empty lines and avoid the issue.

**Comment by eduardomozart** [comment]
That's strange, it shouldn't really happens, maybe DokuWiki main developer "splitbrain" could check if further, because anyone that is drawing their first table could find this issue and think they are making mistake on syntax when they didn't.

**Comment by Klap-in** [comment]
A related, but probably not the same, issue might be lists and tables directly after each other. I saw it mentioned here: https://www.dokuwiki.org/devel:parser#linefeed_grabbing (which also mentions https://bugs.dokuwiki.org/261.html, but that look different again)

See here for a demo page: https://www.dokuwiki.org/sandbox:issue-4051

The unit tests [parser_table.test.php](https://github.com/dokuwiki/dokuwiki/blob/master/_test/tests/inc/parser/parser_table.test.php) cover the case with directly writing normal text before and after tables. Probably it gets tricky when other syntaxes are written directly before and after a table. Some other tests can be added (you are welcome to contribute these :) ) which cover other combinations with table syntax. Having tests that first fail and probably show already (partly) where things are misinterpreted, are a nice starting point.

**Comment by fiwswe** [comment]
Another odd related issue:

```
| 1 |
  
| 2 |
```
(Line with two spaces between the tables.)

Renders as a single table with 3 columns. Is this intentional?

**Comment by tormec** [comment]
The [documentation](https://www.dokuwiki.org/wiki:syntax#basic_text_formatting) says:
> Paragraphs are created from blank lines.

**Comment by fiwswe** [comment]
@tormec Not sure I see what you are getting at? We are talking about lines that start with a certain amount of white space, not blank lines.

**Comment by eduardomozart** [comment]
> @tormec Not sure I see what you are getting at? We are talking about lines that start with a certain amount of white space, not blank lines.

I believe that he wanted to say that tables ends with a blank line and if there's no blank line the table didn't render as expected.

**Comment by fiwswe** [comment]
Well yes, sure. A blank line will terminate the table just fine. And that is certainly a workaround for the issues.

But that does not explain the weird rendering when a line starting with 2+ characters of whitespace is used in a table? Nor does it explain why lines starting with white space are part of the table syntax?

Or to put it differently: So yeah, someone built this nice parsing code for lines starting with 2+ white spaces but don't use it because weird things happen if you do.

**Comment by eduardomozart** [comment]
Maybe those two extra spaces are being rendered as lists? Because lists use two spaces for their rendering.

**Comment by fiwswe** [comment]
No, they render as an extra column, see [above](https://github.com/dokuwiki/dokuwiki/issues/4051#issuecomment-1708229410). But in the case where they are followed by a non-table element, they don't do that.

Maybe @splitbrain can explain what those lines are supposed to do as he [committed](https://github.com/dokuwiki/dokuwiki/commit/36dc94bb8b05aaaff6fdcf55dd6af80ca30d22b1) the code? For all I can see, we could delete https://github.com/dokuwiki/dokuwiki/blob/bac51761dc7fec256b7e3dbfe1205fbf1aad9a16/inc/Parsing/ParserMode/Table.php#L35

**Comment by tormec** [comment]
The [documentation](https://www.dokuwiki.org/wiki:syntax#tables) also says:
> Just add at least two whitespaces at the opposite end of your text: Add two spaces on the left to align right, two spaces on the right to align left and two spaces at least at both ends for centered text.

To be honest I don't see any issue or weird rendering by DokuWiki.

This is my understanding:
1. to start a new paragraph the previous line must be blank[*] (this answers the original issue);
2. 2+ white spaces in a cell without text don't terminate the table (because of 1.);
3. 2+ white spaces before or after a text in a table cell are used to align the text in the cell.
  
[*] Actually, a new paragraph can start also if the previous line contains only a single white space (but I think it's negligible).

**Comment by Klap-in** [comment]
Matching multiple repeating spaces/tabs is done to determine the alignment. See the Doku_Handler, were these matches are converted in calls 
https://github.com/dokuwiki/dokuwiki/blob/bac51761dc7fec256b7e3dbfe1205fbf1aad9a16/inc/parser/handler.php#L1045-L1048

**Comment by splitbrain** [comment]
just a few comments from my phone...

first of all alignment *inside* the cells should not have anything to do with the issue

the issue is where different parser modes that rely on newlines for entry/exit patterns meet each other directly

what one mode consumes is invisible for any other mode. 

in OPs case two spaces open a code block and consume the newline for the follow up table to be recognized 

what might work is to use lookahead/lookbehind patterns to match newlines without consuming them

if anyone wants to try, start with unit tests! no parser changes will be merged without unit tests.

**Comment by fiwswe** [comment]
This seems to be partially fixed in Mort. (Partially because a naïve reader would interpret the source as two separate tables with one column and 1 row each because the two spaces are not readily visible.)

```
| 1 |
  
| 2 |
```
(Line with two spaces between the tables.)

Now renders as a single table with 1 column and 2 rows. Effectively the intermediate line with the two spaces is ignored.

Release 2026-07-14b "Mort"
sprintdoc dw
PHP 8.5.9 fpm-fcgi
OpenBSD 7.9
